activerecord/lib/active_record/enum.rb

```diff
@@ -107,11 +106,15 @@ def type_cast_from_user(value)
 
       def type_cast_from_database(value)
         return if value.nil?
-        mapping.key(value.to_i)
+        mapping.key(super)
       end
 
       def type_cast_for_database(value)
-        mapping.fetch(value, value)
+        super mapping.fetch(value, value)
+      end
+
+      def changed?(old_value, new_value, _new_value_before_type_cast)
+        old_value != new_value
```

Comment:
Ah, I see. Numeric's definition won't like this. The fact that we had to do this bugs me, but I don't have a better solution.